### ISSUE TITLE
residue_constraints ignored

### ISSUE THREAD

**Issue by sven-lange** [HannesStark/boltzgen#247]
Hi,

I have been playing around with residue_constraints to avoid the ubiquitin sequence generation in binders, through forcing certain amino acids (W, C) that are absent in ubiquitin, or including dipeptides that are absent in ubiquitin. But including this constraint in the configuration seems to be ignored entirely and none of the generated binders contain these amino acids in the specified positions. I also didn't see the "residue_constraints" being used in any of the examples apart from in examples/design_spec_showcasing_all_functionalities.yaml 

```
entities:
  - protein:
      id: A
      sequence: 41..90
      residue_constraints:
        - position: 8
          allowed: W
        - position: 20
          allowed: N
        - position: 21
          allowed: K
        - position: 40
          allowed: S
        - position: 41
          allowed: D

**Comment by buerbaumer** [comment]
Hi @sven-lange, 
thanks for reporting this and for providing the YAML snippet.

I checked your example against v0.3.2/current main, and the syntax is valid. Positions are 1-indexed relative to the protein entity, so your configuration should force chain A to contain:

W at position 8
N at position 20
K at position 21
S at position 40
D at position 41

These are intended to be hard constraints during inverse folding: all other amino-acid logits at those positions are masked. Therefore, an inverse-folded sequence should always satisfy them.

A few situations could make the constraints appear to be ignored:

1.) The constraints apply only during inverse folding.
Files under intermediate_designs/ are pre–inverse-folding backbone designs and do not represent the constrained final sequence. Please inspect intermediate_designs_inverse_folded/ or the final ranked outputs.
Also ensure that the run did not use --skip_inverse_folding. If --steps was supplied, it must include inverse_folding.

2.) Previously generated outputs may be reused.
If the YAML was changed while running against an existing output directory with --reuse, already generated and inverse-folded designs are skipped. Please test with a new output directory and without --reuse.

3.) Please confirm the installed version.
boltzgen --version
residue_constraints was introduced in v0.3.1. The current PyPI release is v0.3.2.

4.) Cysteine can conflict with the protocol default.
The peptide, antibody, and nanobody protocols exclude C by default through --inverse_fold_avoid C. A per-position allowed: C conflicts with that global exclusion, and the global exclusion currently takes priority.
To permit or force cysteine with those protocols, explicitly clear the default:
--inverse_fold_avoid ""



Could you please try a clean minimal run?
boltzgen check issue247.yaml

boltzgen run issue247.yaml \
  --output issue247-clean \
  --protocol protein-anything \
  --steps design inverse_folding \
  --num_designs 5

Then inspect the chain-A sequences in:
issue247-clean/intermediate_designs_inverse_folded/
For the whitelist constraints in your example, five designs are sufficient for verification: every generated sequence should contain W, N, K, S, and D at positions 8, 20, 21, 40, and 41 respectively.

If the clean run still violates the constraints, could you provide:
the output of boltzgen --version
the complete command and protocol
the complete design YAML
whether --reuse, --skip_inverse_folding, symmetric_group, or a custom --inverse_fold_avoid was used
any warnings mentioning aa_constraint_mask, shape mismatches, or conflicting restrictions
one corresponding .cif and .npz file from both intermediate_designs/ and intermediate_designs_inverse_folded/ (zipped if GitHub does not accept the extensions)

Regards,
Harald
